app/services/post_alerter.rb

```diff
@@ -274,11 +274,15 @@ def category_or_tag_muters(topic)
       SELECT user_id FROM tag_users tu JOIN topic_tags tt ON tt.tag_id = tu.tag_id AND tt.topic_id = #{topic.id} AND tu.notification_level = #{TagUser.notification_levels[:muted]}
     SQL
     User
-      .where("id IN (#{user_ids_sql})")
+      .where("users.id IN (#{user_ids_sql})")
       .joins("LEFT JOIN user_options ON user_options.user_id = users.id")
+      .joins("LEFT JOIN topic_users tu ON tu.user_id = users.id AND tu.topic_id = #{topic.id}")
       .where(
         "user_options.watched_precedence_over_muted IS false OR (user_options.watched_precedence_over_muted IS NULL AND #{!SiteSetting.watched_precedence_over_muted})",
       )
+      .where(
+        "tu.id IS NULL OR tu.notification_level != #{TopicUser.notification_levels[:watching]}",
```

Comment:
I think we might be able to put this in the `user_ids` subquery instead of joining the users table against the `topic_users` table.